[Tools] Log analyzer for distributed

pbench opened the pull request:
The log analyzer produces and Gantt chart of the different computations performed during a "journey" call to Navitia.
It does so by parsing the logs of Jormungandr and Kraken.
It is especially useful in the "distributed" context, to see which part of the computations take a long time.

For each request to Navitia, a `request_id` is generated in `jormungandr/interfaces/v1/Journeys.py:get:generate_request_id`.
This identifier is kept around all computations :
 - it is used in Jormungandr each time the `timed_logger` of `jormungandr/scenarios/helper_classes/helper_utils.py`
 - it is sent to Kraken via the `request_id` field of the protobuf, and used in `kraken/kraken_zmq.h:do_work`

The `log_analyzer` identify the log lines corresponding to computation timing, parses them, group them by request, and then produces a chart for each request.


See `source/log_analyzer/readme.md` for usage instructions

Commit:
generate an id for a request based on the input parameters

Commit:
adding request_id in distributed and new_default

Commit:
fix after tests failed

xlqian commented on source/jormungandr/jormungandr/interfaces/v1/Coverage.py:
```diff
@@ -59,7 +60,8 @@ def __init__(self, quota=True, *args, **kwargs):
         args = self.parsers["get"].parse_args()
 
-        resp = i_manager.regions(region, lon, lat)
+        request_id = "coverage_{}".format(flask.request.id)
```
you really need this one for distributed?

pbench replied:
Not really, I was just observing `Coverage` requests in kraken, and wondering where they come from.

xlqian commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -485,9 +485,45 @@ def _set_specific_params(mod):
+            return result
+
+        request_id = generate_request_id()
+        args["request_id"] = request_id
```
flask.request.id is not enough?

xlqian replied:
OK, you want to be able to compare the response time between two scenarios right?

pbench replied:
> OK, you want to be able to compare the response time between two scenarios right?

Exactly, since the changes in response time between new_default and distributed seems to vary with the request, I thought it may be useful to have the same request_id for the same request (disregarding the scenario used, and the datetime at which the request was made)
Maybe an overkill though

Commit:
fix tests after rebase

Commit:
make pre-commit happy

Commit:
- add suffix to request_id in subrequests
- propagate request_id to jomun timed_logger
- add ids to kraken workers in logs

Commit:
fix mock_kraken compilation

kadhikari commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -488,11 +488,53 @@ def _set_specific_params(mod):
+
+            import uuid
+
+            result = str(uuid.uuid4())
```
Why we don't use the value of result calculated above ?

pbench replied:
thanks, I don't know why I added this at some point

kadhikari commented on source/jormungandr/jormungandr/scenarios/helper_classes/proximities_by_crowfly.py:
```diff
@@ -51,18 +59,19 @@ def __init__(
+            self._request_id,
+            **self._speed_switcher
+        )
 
```
Don't we need 'timed_logger' here ?

pbench replied:
you're right ! Thanks

kadhikari commented on source/jormungandr/jormungandr/scenarios/helper_classes/streetnetwork_path.py:
```diff
@@ -73,24 +74,26 @@ def __init__(
+        except Exception as e:
+            logging.getLogger(__name__).error("Exception':{}\n".format(str(e)))
+
+            return None
```
Same here 'timed_logger' removed why?

pbench replied:
you're right ! Thanks

Commit:
fix after review + subrequest name for origin/dest places and car_co2

Commit:
analyze kraken and jormun log to produce gantt chart

pbench commented:
I added the code to generate the gantt chart from kraken and jormun logs in source/log_analyzer. Maybe another repo is more suited, I don't know.

Commit:
report jormun logger name in gantt

Commit:
documentation

Commit:
test without debian9

Commit:
test only one debian

Commit:
putting back all 3 debian

nonifier commented on source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py:
```diff
@@ -468,11 +477,19 @@ def check_final_results_or_raise(final_results, orig_fallback_durations_pool, de
     finally:
-        collapsed_time = time.time() - start
+        end = time.time()
+        collapsed_time = end - start
```
I think it's meant to be `elapsed_time`. Unless you expect the time to collapse on us, but you might not want that :D

nonifier commented on source/kraken/kraken_zmq.h:
```diff
@@ -105,7 +107,10 @@ inline void doWork(zmq::context_t& context,
 
         api = pb_req.requested_api();
-        log4cplus::NDCContextCreator ndc(pb_req.request_id());
+        std::string request_id = pb_req.request_id();
```
```suggestion
        const std::string& request_id = pb_req.request_id();
```
Unless you really want to work on a copy ?

nonifier commented on source/log_analyzer/log_analyzer.py:
```diff
@@ -0,0 +1,182 @@
+        import colorsys
+
+        nb_of_colors = len(df)
+        HSV_tuples = [(x * 1.0 / nb_of_colors, 0.5, 0.5) for x in range(nb_of_colors)]
```
clever usage of the HSV color space :+1:

pbench replied:
Stackoverflow often give good solutions !
https://stackoverflow.com/questions/876853/generating-color-ranges-in-python

nonifier commented on source/log_analyzer/log_analyzer.py:
```diff
@@ -0,0 +1,182 @@
+        import plotly.figure_factory
+
+        sorted_jormun_tasks = sorted(self.jormun_tasks, key=lambda task: (task.start, -task.end))
+        df = [
+            dict(
+                Task="J{}#{}".format(task.sub_request_id, task.task_name),
+                Start=task.start,
+                Finish=task.end,
+                SubRequest="J{}#{}".format(task.sub_request_id, task.task_name),
+            )
+            for task in sorted_jormun_tasks
+        ]
+        for worker, tasks in self.kraken_tasks_by_worker.items():
+            for task in tasks:
+                df.append(
+                    dict(
+                        Task="Kraken_{}".format(task.worker),
+                        Start=task.start,
+                        Finish=task.end,
+                        SubRequest="K{}".format(task.sub_request_id),
+                    )
+                )
```
just for the sake of pythoness and a bit of fun, this could be refacto a little:

```py
def make_df(task):
   return dict(   Task="Kraken_{}".format(task.worker),
                        Start=task.start,
                        Finish=task.end,
                        SubRequest="K{}".format(task.sub_request_id))

f = [ make_df(task) for task in sorted_jormun_tasks ]
f.extends([ make_df(t) for tasks in self.kraken_tasks_by_worker.values() for t in tasks ])
```

nonifier commented on source/log_analyzer/log_analyzer.py:
```diff
@@ -0,0 +1,182 @@
+        # https://plotly.com/python/gantt/
+        import plotly.figure_factory
+
+        sorted_jormun_tasks = sorted(self.jormun_tasks, key=lambda task: (task.start, -task.end))
```
I don't know if this could be useful here, but if you want to word with sorted list, dict or set, you could have a look at `SortedContainer`. You'll find a bunch of useful data structures.
http://www.grantjenks.com/docs/sortedcontainers/

Commit:
taking into account reviews

Commit:
wording

mbounabi commented on source/jormungandr/jormungandr/instance.py:
```diff
@@ -559,13 +559,11 @@ def _send_and_receive(
         request.deadline = deadline.strftime('%Y%m%dT%H%M%S,%f')
         with self.socket(self.context) as socket:
-            try:
-                request.request_id = flask.request.id
```
Does this change mean that the flask request id was useless?

pbench replied:
No it wasn't, I added it back.

mbounabi commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -30,7 +30,7 @@
 
 from __future__ import absolute_import, print_function, unicode_literals, division
-import logging
+import logging, flask, json, hashlib, datetime
 from flask import request, g
```
Why import flask? Only flask.request seems to be used and it's already imported.

pbench replied:
:+1: removed

mbounabi commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -488,11 +489,50 @@ def _set_specific_params(mod):
         if len(possible_regions) > 1:
             base_args = deepcopy(args)
 
+        logger.debug("We are about to ask journeys on regions : {}".format(possible_regions))
```
There's a similar log here:
https://github.com/CanalTP/navitia/blob/ed9a2b2cccbc503b47d4d13c2a9005bba8e59678/source/jormungandr/jormungandr/interfaces/v1/Journeys.py#L400
Isn't it enough?

pbench replied:
well spotted ! Removed

mbounabi commented on source/log_analyzer/readme.md:
```diff
@@ -0,0 +1,48 @@
+# Log Analyzer
+
+The log analyzer produces and Gantt chart of the different computations performed during a "journey" call to Navitia.
```
```suggestion
The log analyzer produces a Gantt chart of the different computations performed during a "journey" call to Navitia.
```

mbounabi commented on source/log_analyzer/readme.md:
````diff
@@ -0,0 +1,48 @@
+
+```
+
+- prepare the needed python environement with `virtualenvwrapper` (see https://virtualenvwrapper.readthedocs.io/en/latest/ for installation instructions).
````
```suggestion
- prepare the needed python environment with `virtualenvwrapper` (see https://virtualenvwrapper.readthedocs.io/en/latest/ for installation instructions).
```

mbounabi commented on source/log_analyzer/readme.md:
````diff
@@ -0,0 +1,48 @@
+mkvirtualenv -p python3 -r requirements.txt log_analyzer
+```
+
+- then load the python environement, and launch the tool
````
```suggestion
- then load the python environment, and launch the tool
```

mbounabi commented on source/log_analyzer/readme.md:
````diff
@@ -0,0 +1,48 @@
+
+```
+
+This will create in `my/output/dir/` and html file containing the gantt chart for each journey request found in the logs of Jormungandr and Kraken.
````
```suggestion
This will create in `my/output/dir/` an html file containing the gantt chart for each journey request found in the logs of Jormungandr and Kraken.
```

mbounabi commented on source/navitiacommon/navitiacommon/default_values.py:
```diff
@@ -157,12 +157,12 @@ def get_value_or_default(attr, instance, instance_name):
-                value,
-                attr,
-            )
+            # logger.warning(
+            #     'instance %s not found in db, we use the default value (%s) for the param %s',
+            #     instance_name,
+            #     value,
+            #     attr,
+            # )
```
Why not simply delete it?

pbench replied:
I put it back. I think it was annoying me when looking at logs when developing.

Commit:
fix after review

Commit:
fix for python3

xlqian commented on source/log_analyzer/log_analyzer.py:
```diff
@@ -0,0 +1,189 @@
+    def add_kraken_task(self, api, sub_request_id, worker, start, end):
+        worker_tasks = self.kraken_tasks_by_worker.get(worker, [])
+        worker_tasks.append(KrakenTask(api, sub_request_id, worker, start, end))
+        self.kraken_tasks_by_worker[worker] = worker_tasks
```
this is what you want to use i think: https://docs.python.org/3.8/library/collections.html#defaultdict-examples

pbench replied:
yes ! Thanks, I didn't know about it

xlqian reviewed (approved):
nice tool!!!!

Commit:
makes pre-commit happy
